src/doorman/Fd.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "Byte.hpp"
+
+#include <gsl/span>
+
+#include <type_traits>
+#include <utility>
+
+#include <sys/socket.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+class Fd {
+    int fd_;
+
+    void writev(gsl::span<const iovec> io_vecs) const;
+
+    static void *iovec_cast(const void *thing) { return const_cast<void *>(thing); }
+
+    template <typename T>
+    static iovec iovec_for(const T &some_object) {
+        static_assert(std::is_trivially_copyable_v<T>);
+        return {iovec_cast(&some_object), sizeof(some_object)};
+    }
+
+    static iovec iovec_for(gsl::span<const byte> bytes) { return {iovec_cast(bytes.data()), bytes.size_bytes()}; }
+    static iovec iovec_for(std::string_view string) { return {iovec_cast(string.data()), string.size()}; }
+
+    template <typename... Args>
+    static std::array<iovec, sizeof...(Args)> iovecs_for(Args &&... args) {
+        return {iovec_for(args)...};
+    }
+    void read_all(void *data, size_t length) const;
+    size_t try_read_some(void *data, size_t length) const;
+
+public:
+    Fd() : fd_(-1) {}
+    explicit Fd(int fd) noexcept : fd_(fd) {}
+    ~Fd() noexcept { close(); }
+    Fd(const Fd &) = delete;
+    Fd &operator=(const Fd &) = delete;
+    Fd(Fd &&other) noexcept : fd_(std::exchange(other.fd_, -1)) {}
+    Fd &operator=(Fd &&other) noexcept {
+        close();
+        fd_ = std::exchange(other.fd_, -1);
+        return *this;
+    }
+    [[nodiscard]] bool is_open() const noexcept { return fd_ >= 0; }
+    [[nodiscard]] int number() const {
+        if (!is_open())
+            throw std::runtime_error("Attempt to get handle for invalid file descriptor");
+        return fd_;
+    }
+    void close() {
+        if (is_open())
+            ::close(fd_);
+        fd_ = -1;
+    }
+    void write(gsl::span<const char> span) const { write(span.data(), span.size_bytes()); }
+    void write(gsl::span<const byte> span) const { write(span.data(), span.size_bytes()); }
+    void write(const void *data, size_t length) const;
+    template <typename T>
+    void write(const T &t) const {
+        static_assert(std::is_trivially_copyable_v<T>);
+        write(&t, sizeof(t));
+    }
+    template <typename... Args>
+    void write_many(Args &&... args) const {
+        writev(iovecs_for(std::forward<Args>(args)...));
+    }
+
+    template <typename T>
+    T read_all() const {
```

Comment:
It looks like there's one place that calls this version: Xania.cpp:84, to read a Packet. And you would expect the compiler to use return value optimization in that case?

Reply:
Absolutely! Named return value optimization (nRVO): https://godbolt.org/z/5YzMYa you'll see that the call to `read_all<BigThing>` on line 11 sets up `rdi` to point on the stack. `rdi` is the first param...the example function takes none! The compiler behind the scenes has rewritten it to be:

```
template <typename T>
[[gnu::noinline]] T read_all(T &a_t_in_the_calling_context) {
  read_all(&a_t_in_the_calling_context, sizeof(a_t_in_the_calling_context));
}
```